src/login/bpf/logind-revoke.bpf.c

```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+/* The SPDX header above is actually correct in claiming this was
+ * LGPL-2.1-or-later, because it is. Since the kernel doesn't consider that
+ * compatible with GPL we will claim this to be GPL however, which should be
+ * fine given that LGPL-2.1-or-later downgrades to GPL if needed.
+ */
+
+#include <linux/types.h>
+#include <linux/bpf.h>
+#include <linux/hidraw.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+extern int hid_bpf_hidraw_revoke(int fd) __ksym;
+
+struct hidraw_revoke_syscall_args {
+       int fd;
+};
+
+SEC("syscall")
+int hidraw_revoke(struct hidraw_revoke_syscall_args *args)
+{
+       return hid_bpf_hidraw_revoke(args->fd);
+}
```

Comment:
even in bpf code we generally try to follow our coding style, i.e. `{` on same line as function name, but doesn't really matter

Reply:
btw, what's the rationale of exposing this via ebpf? i mean, the call is simply enough to just be an ioctl, no?

sounds like quite some heavy machinery for something that could just be an ioctl? can you explain the background of this?

Reply:
> sounds like quite some heavy machinery for something that could just be an ioctl? can you explain the background of this?

Completely agree, and TBH I can not convince myself  it's the right way to go (which also explain why this wasn ´t sent to the LKML). So dumping my thoughts:

On the "why should we use eBPF" we have:
- eBPF kfuncs are considered as kABI, in the same way `EXPORT_GPL_SYMBOL` are. Which means that if we get the API wrong, we can always change it, it's kernel-to-kernel API
- eBPF kfuncs are validated by the eBPF verifier. So if we decide to change the API because of the previous point, we can guarantee that logind will be using the proper behavior
- we need eBPF for firewalling devices in the long run, so having the mechanics in place now kind of makes sense
- we can skip the fight for a generic "revoke" approach on the ioctl front, because it's going to be very hard to work.
- we could use a similar eBPF kfunc to revoke a fd owned by a different process. We can then add a "monitoring" eBPF program with logind that would also check for any other opened hidraw nodes, and include them in the revoke part when we need to do it.

On the "why using eBPF is pointless":
- the new ioctl is simple enough an known enough (but is uapi, so can not easily been changed)
- it's a heavy machinery to just get a single syscall
- the  "let's revoke other processes' fd" use case is not actually implemented, and might never be

So from the kernel side, I think the main advantage to using eBPF here is to not tie ourselves with a given API, meaning that we can extend/restrict it further without having to worry too much.

Opinions/thoughts?

Reply:
> * eBPF kfuncs are considered as kABI, in the same way `EXPORT_GPL_SYMBOL` are. Which means that if we get the API wrong, we can always change it, it's kernel-to-kernel API

Are you sure Linus really sees it that way? If we start using this symbol from userspace for things that are not purely tracing, are you sure you can just take it away again?

I mean, I can see that this is politically easier, I sympathize with that, but as I understand Linux once it is loadbearing for userspace it's not that easy to remove again.

Reply:
but frankly, I am fine with either. You have to fight the political fights on the kernel side. If this makes it simpler for you we are fine with the bpf approach. we are all in on bpf, no issues there.

Reply:
> Are you sure Linus really sees it that way? If we start using this symbol from userspace for things that are not purely tracing, are you sure you can just take it away again?

I sneaked in last year kernel summit when this very same topic was discussed (and for that reason). His answer was definitely that eBPF is kernel-to-kernel, and that the only users of it are either big corporations running this on millions of machines, or systemd, in which case it's not a "normal" user, as they can know better on how to handle the situation.

It's even explicitly mentioned on the kernel docs:
https://docs.kernel.org/bpf/kfuncs.html#bpf-kfunc-lifecycle-expectations

> kfuncs provide a kernel <-> kernel API, and thus are not bound by any of the strict stability restrictions associated with kernel <-> user UAPIs. This means they can be thought of as similar to EXPORT_SYMBOL_GPL, and can therefore be modified or removed by a maintainer of the subsystem they're defined in when it's deemed necessary.
>
> Like any other change to the kernel, maintainers will not change or remove a kfunc without having a reasonable justification.

So basically, as I see it, if we agree for using a kfunc as a system call, logind should have a way to "test" if the kfunc is available, by attempting to load it once, and keep a boolean somewhere. This way, if that kfunc were to change, logind wouldn't break, the feature would not be available.

I think the easiest would be to attach the revoke BPF program only once, when the first a user of the hidraw fd appears, and only remove it when the last user of the last hidraw fd vanished. This should remove part of the heaviness by loading the bpf program once, and we can even try different versions of it, without doing such dance every time we need to revoke a hidraw fd.

Reply:
Just chiming in with a summary of yesterday's chat with @bentiss:

`HIDIOCREVOKE` has a very specific use-case that we know (because we've done it with `EVIOCREVOKE`) and it's a tiny kernel patch and trivial to maintain.

BPF-based revoke can give us the same result with more effort and it has much greater flexibility. Alas, the use-cases we have for that greater flexibility are mostly theoretical, we don't know if we need them (things like revoking someone else's fd) and/or how we'd get to the point where we definitely need those. Other things we know we'll need BPF for (firewalling parts of the device) are orthogonal to revoke.

So right now the pendulum of motivation is swinging back to the ioctl :)

Reply:
@whot after making significant progress in the USB portal (and understanding the problem domain a bit more to believe I can provide feedback on this HID review PR), I think we'll need a way to revoke device access from other apps.

Reply:
@GeorgesStavracas - just to confirm, you need to be able to revoke an fd that you didn't open and pass along (like logind does)? Can you provide a few more details on the use-case here please?

Reply:
Right now logind only opens once (see [method_take_device](https://github.com/systemd/systemd/blob/main/src/login/logind-session-dbus.c#L482)). That makes sense for evdev (well, mice/keyboards) but not so much for hidraw so I'd say we need to extend the logind API to be able to re-open a device so they can be individually revoked.